webrtc/RTCDataChannel-send-receive-stress.html

```diff
@@ -0,0 +1,176 @@
+    const view = new DataView(message.buffer);
+    view.setUint16(0, index);
+    if (index % 3 === 1) {
+      return [message, message];
+    } else {
+      // TODO: Reactivate this once Chrome has Blob support :|
+      // return [new Blob([message]), message];
+      return [message, message];
+    }
+  });
+
+  // Create channel pairs
+  const nChannelPairs = channelOptionsVariants.length * 30;
+  const dcPairs = await Promise.all(Array.from(new Array(nChannelPairs), (_, i) => {
+    let [labelFragment, channelOptions] = channelOptionsVariants[i % channelOptionsVariants.length];
+
+    // Modify ID (if negotiated)
+    if (channelOptions instanceof Array) {
+      channelOptions = channelOptions.map((options) => {
+        // Note: Adding '100' here to avoid clashes with in-band negotiated channels
+        options.id = i + 100;
+        return options;
+      });
+    }
+
+    // Create pair
+    return createDataChannelPair(pc1, pc2, {
+      channelLabel: `${labelFragment}, index=${i}`,
+      doSignaling: i === nChannelPairs - 1,
+      channelOptions: channelOptions
+    });
+  }));
+
+  // Received message indexes per channel per peer
+  const indexesArray = [];
+
+  for (let i = 0; i < dcPairs.length; ++i) {
+    for (const dc of dcPairs[i]) {
+      const indexes = new UniqueSet();
+      indexesArray.push(indexes);
+      dc.binaryType = 'arraybuffer';
+
+      // Verify the data on receive
+      dc.onmessage = t.step_func(event => {
+        const {data} = event;
+
+        // String: Split by semicolon and compare string
+        let index;
+        if (!(data instanceof ArrayBuffer)) {
+          index = data.split(';')[0];
+          index = parseInt(index, 10);
+          const [, expected] = messages[index];
+          assert_equals(data, expected,
+            'Received data should be equal to sent data');
+        } else {
+          const view = new DataView(data);
+          index = view.getUint16(0);
+          const [, expected] = messages[index];
+          assert_equals_typed_array(data, expected,
+            'Received data should be equal to sent data');
+        }
+
+        // Add received message index (will check the index is unique)
+        indexes.add(index);
+
+        // Done if all messages on all channels on both peers have been received
+        if (index === messages.length - 1
+            && indexesArray.every((indexes) => indexes.size === messages.length)) {
+          t.done();
+        }
+      });
+    }
+  }
+
+  for (let i = 0; i < dcPairs.length; ++i) {
+    for (const dc of dcPairs[i]) {
+      // Delay sending
+      t.step_timeout(() => {
+        // Start transmitting messages
+        for (const [message] of messages) {
+          dc.send(message);
+        }
+      }, delay * (i + 1));
+    }
+  }
+
+  await t.done_promise;
+}, 'Stress-test with multiple channels sending and receiving using various data sources', {
+  timeout: 420000 // yep, 7m to be safe
```

Comment:
I would suggest leaving the long-duration stress test in its own file, to minimize the use of explicit timeouts. It's the only place in the CL with a timeout that is longer than the default.